fmpz/mul_2exp.c

```diff
@@ -9,38 +10,103 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= FLINT_BITS - 2)  /* Result fits inside a small fmpz */
```

Comment:
Problem here: if c1bits + exp overflows, we can enter this branch even if c1bits > FLINT_BITS - 2, and then we end up dereferencing an invalid pointer. We probably want to handle overflow correctly in this method, even when the exponent is so large that the result can't be stored in memory.

Reply:
Thanks for your input. Would it be okay to say this in its documentation, rather than slowing the function down?